text/0008-kadalu-kubernetes-storage-monitoring.adoc

```diff
@@ -0,0 +1,153 @@
+- How quick a PV is provisioned?
+
+==== Events
+
+A few Events can be derived from the collected metrics by comparing with the latest data with the previously collected metrics. For example,
+
+- *Restarted Event* - If the current uptime of a resource is less than the previous one.
+- *Upgraded Event* - If the current version is greater than the previous one.
+- *Option Change* - If the list of Volume options is different than the previously collected options data.
+- *New Storage added* - When the latest list of metrics shows a new Storage compared to previously collected data.
+
+These events will be exposed by `GET /events` URL. These events also helps to understand the storage better when viewed along with the historical metrics. For example, if a Performance degradation is observed in the chart and associated event says that a Option is changed or one Replica server Pod was down or the Cluster is upgraded to a new Version.
+
+==== Alerts
+
+All events may not be useful, some events may be very noisy. For example, an Event related to changes to utilization is very noisy. But becomes very important when it meets some criteria.
+
+- When utilization crosses 70%.
+- Performance improvement/degrade.
+- Storage is down.
+
+=== List of Metrics
+
+- *Number of Storage pools*
+- *Number of PVs*
+- *Number of Storage Units/Bricks*
+- *Operator Health* - Operator is running or not
+- *Storage Units Health* - Health of Brick processes, `1 for Up, and 0 for Down`
+- *Self Heal daemon Health*
+- *Health of Metrics exporter*
+- *CSI Provisioner Health*
+- *CSI/Quotad health*
+- *CSI/Mounts health* (Based on expected number of Volumes in ConfigMap and number of mount processes). Gluster client process will continue to run even if all the bricks are down, it waits for the brick processes and re-connects as soon as they are available. Detect this by doing a regular IO from the mount or parsing the log files for `ENOTCONN` errors.
+- *CSI/Storage and PV utilization*
+- *Storage Units Utilization*
+- *Storage units/bricks CPU,Memory and Uptime metrics*
+- *CSI Provisioner CPU,Memory and Uptime metrics*
+- *CSI Node plugins CPU,Memory and Uptime metrics*
+- *Heal Pending metrics* - Run `glfsheal` command and get the heal pending count per Volume.
+- *CSI/Node plugin mounts health*
+
+Advanced metrics can be introduced in the future versions
+
+- *Performance metrics* - Using Gluster Volume profile and other tools available.
+
+=== Implementation
+
+Every Kadalu container(Including the Operator) will have a HTTP server process that exposes one API `/_api/metrics`. This API need not be exposed outside the Cluster. Only Operator needs access to this API.
+
+Operator container will have a HTTP server process that exposes following two new APIs along with `/_api/metrics`.
+
+----
+GET /metrics
+GET /metrics.json
+----
+
+Prometheus annotations required only for Operator Pod yaml file as follows.
+
+[source,yaml]
+----
+      annotations:
```

Comment:
- Scrape interval is configurable by user, like another annotation would suffice here?

Reply:
Prometheus is Pull based, that means it calls the APIs and collects the metrics. Metric exporters should not have its own scrape interval https://prometheus.io/docs/instrumenting/writing_exporters/#scheduling